src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/services/LinkedStudyServiceImpl.java

```diff
@@ -0,0 +1,160 @@
+import uk.ac.ebi.ampt2d.metadata.persistence.repositories.StudyRepository;
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.Stack;
+import java.util.stream.Collectors;
+
+
+public class LinkedStudyServiceImpl implements LinkedStudyService {
+
+    @Autowired
+    private StudyRepository studyRepository;
+
+    @Autowired
+    private LinkedStudyRepository linkedStudyRepository;
+
+    private List<LinkedStudy> findLinkedStudies(AccessionVersionEntityId id) {
+        QLinkedStudy linkedStudy = QLinkedStudy.linkedStudy1;
+        Predicate predicate = linkedStudy.study.id.eq(id).or(linkedStudy.linkedStudy.id.eq(id));
+
+        return (List<LinkedStudy>) linkedStudyRepository.findAll(predicate);
+    }
+
+    private List<AccessionVersionEntityId> findLinkedStudyIds(AccessionVersionEntityId id) {
+        Set<AccessionVersionEntityId> ids = new HashSet<>();
+        Set<AccessionVersionEntityId> seen = new HashSet<>();
+        Stack<AccessionVersionEntityId> stack = new Stack<>();
+
+        seen.add(id);
+        stack.push(id);
+
+        while ( !stack.empty() ) {
+            AccessionVersionEntityId top = stack.pop();
+            List<LinkedStudy> linkedStudies = findLinkedStudies(top);
+
+            for ( LinkedStudy linkedStudy : linkedStudies ) {
+                Arrays.asList(linkedStudy.getStudy().getId(), linkedStudy.getLinkedStudy().getId())
+                        .forEach(studyId->{
+                            if ( !top.equals(studyId) && !seen.contains(studyId) ) {
+                                ids.add(studyId);
+                                seen.add(studyId);
+                                stack.push(studyId);
+                            }
+                        });
+            }
+        }
+
+        return ids.stream().collect(Collectors.toList());
+    }
+
+    @Override
+    public LinkedStudy findOne(AccessionVersionEntityId studyId, AccessionVersionEntityId linkedStudyId) {
+        Study study = studyRepository.findOne(studyId);
+        Study linkedStudy = studyRepository.findOne(linkedStudyId);
+
+        if ( study == null || linkedStudy == null) {
+            return null;
+        }
+
+        LinkedStudyId linkedStudyId1 = new LinkedStudyId(studyId, linkedStudyId);
+        LinkedStudy linkedStudy1 = linkedStudyRepository.findOne(linkedStudyId1);
+
+        if ( linkedStudy1 != null ) {
+            return linkedStudy1;
+        }
+
+        LinkedStudyId linkedStudyId2 = new LinkedStudyId(linkedStudyId, studyId);
+        LinkedStudy linkedStudy2 = linkedStudyRepository.findOne(linkedStudyId2);
+
+        if ( linkedStudy2 != null ) {
+            return linkedStudy2;
+        }
+
+        return null;
+    }
+
+    @Override
+    public List<Study> findLinkedStudiesById(AccessionVersionEntityId id) {
+        List<AccessionVersionEntityId> ids = findLinkedStudyIds(id);
```

Comment:
You can create a method in the repository findByAorB(A,B)

Reply:
In this case we can't. The `findLinkedStudyIds` will find all ids for all the studies that are linked <b>transitively</b>. So `findByAorB(id,id)` does not meet this requirement.
I will add documentation to the method to clarify this point.